docs/src/app/(public)/(content)/react/utils/use-render/page.mdx

```diff
@@ -0,0 +1,89 @@
+# useRender
+
+<Subtitle>Utility for adding Base UI-like features to custom components.</Subtitle>
+<Meta name="description" content="Utility for adding Base UI-like features to custom components." />
+
+The `useRender` hook allows you to support the same features Base UI provides across all components so that you can also have a consistent experience in your custom components:
+
+- A [render](/react/handbook/composition) prop to override the default rendered element (you may find this concept as "slot" in other libraries).
+- A [callback](/react/handbook/styling#css-classes) on the `className` prop that enables passing dynamically generated CSS classes based on the component's state.
+- Adds [data-attributes](https://base-ui.com/react/handbook/styling#data-attributes) that map to the component's state that can be used as style hooks.
+
+## API reference
+
+### Input parameters
+
+<PropsReferenceTable
+  data={{
+    className: {
+      type: 'string | ((state: State) => string)',
+      description:
+        "CSS class applied to the element, or a function that returns a class based on the component's state.",
+    },
+    render: {
+      type: 'RenderProp<State>',
+      description:
+        "Allows you to replace the component's HTML element\nwith a different tag, or compose it with another component.\n\nAccepts a \`ReactElement\` or a function that returns the element to render.",
+    },
+    state: {
+      type: 'State',
+      description:
+        'The state of the component. It will be used as a parameter for the render and className callbacks.',
+    },
+    props: {
+      type: 'Record<string, unknown>',
+      description: 'Props to be spread on the rendered element.',
+    },
+    stateAttributesMap: {
+      type: 'StateDataAttributes<State>',
+      description: 'A mapping of state to style hooks.',
+    },
+  }}
+/>
+
+### Return value
+
+The hook returns a function that when called returns the element that should be rendered.
+
+## Usage
+
+### Supporting the `render` prop
+
+This is an example of a Text component that provides the support for the render prop.
+
+<Demo path="./demos/render" />
+
+### Generating data attributes
```

Comment:
I think we could come up with a better demo. Clicking a text is not very common and it's hard to discover. We could also simplify it and only leave one "state" (color or size, but not both).

We could also use this demo only for data attributes and leave the className callback as paragraph + code snippet in a new section. This could also improve the discoverability of the className callback docs.

Reply:
Yeah, the thing I was trying to add here is that people can use internal state as well, I couldn't come up with an idea for internal state for Text component :) I am ok with simplifying it.

Reply:
I changed the demo to be a counter. Also added a separate demo for the className callback.